GSoC 2025 - Block reading and bicameral macros

Frityet opened the pull request:
Project info: https://summerofcode.withgoogle.com/programs/2025/projects/kiO8b1Ql

hishamhm committed:
begin splitting modules

hishamhm committed:
begin splitting modules

hishamhm committed:
continue splitting modules

hishamhm committed:
continue splitting modules

hishamhm committed:
Makefile: clean target

hishamhm committed:
split AST traversal

hishamhm committed:
split AST traversal

hishamhm committed:
continue conversion into modules

hishamhm committed:
split type errors into module

hishamhm committed:
split teal.environment module

hishamhm committed:
refactor: add default_env function

hishamhm committed:
precompiled default environment

hishamhm committed:
refactor: move fresh_typeargs to teal.types

hishamhm committed:
refactor: move target_from_lua_version to teal.gen.lua_generator

hishamhm committed:
refactor: move default_env to teal.environment

hishamhm committed:
refactor: split type_checker; it needs to be further split

hishamhm committed:
precompiler: generate much smaller output

Shrinks teal/precompiled/default_env.lua
from 189928 to 100747 bytes.

Compilation times for `tl check hello.tl`
(where `hello.tl` is `print("hello")`,
i.e. measuring compiler baseline overhead)

* master: 21.1 ms
* original precompiler: 15.9 ms
* revised precompiler: 13.9 ms

hishamhm committed:
refactor: split more parts from checker

hishamhm committed:
refactor: move Facts metatables to teal.facts

hishamhm committed:
refactor: move fact engine to teal.facts

hishamhm committed:
Makefile: use a _temp dir for temporary files

hishamhm committed:
Makefile: make it more readable again

It was getting unwieldy. Move the more involved logic
to an `extras/make.sh` Bash script, which I tried to make
as readable as possible.

hishamhm committed:
simplify prelude and stdlib generation

They are now plain .d.tl files, and in `new_env` (which is
currently the "manual" environment builder), they are loaded using the
standard require_module mechanism, same as the other predefined_modules
(or any other uses of `require` by Teal code, really!)

Of course, `environment.default` now uses the precompiled one.

And yes, this is an experiment in the direction of perhaps someday
providing a precompiled type definition facility for all modules,
aiming to speed up compilations. This has the difficulty that the
Teal environment currently relies on globally unique numeric typeids.
We'd have to switch to a name-based system with a linking step.
(The standard library can get away with this because it is always
the first module on an empty environment, so I can reliably start
my typeid count from 1.)

Frityet commented:
@hishamhm

upedd committed:
parser: attach comments to AST nodes

See 4ccded20711ba5842541557d46391fd1bd13c1bf
See 4abd8e2758e5ac3861c93779ff91114fed134743

hishamhm committed:
refactor: move all environment generation to teal.environment

hishamhm committed:
refactor: move package loader to teal.package_loader

hishamhm commented:
@Frityet please load commit [586a137](https://github.com/Frityet/tl/pull/1/commits/586a137467c6f69cc787bc07aa316cf89429eed7) as the `module-split` branch in this repo, and then retarget this PR not to `master`, but to that instead. This way only your commits will show up in this PR!

```
git checkout -b module-split
git reset --hard 586a137
git push
```

You can retarget the PR by clicking the "Edit" button that should appear for you at the top right, next to "<> Code"

hishamhm commented on teal/block-parser.tl:
```diff
@@ -0,0 +1,638 @@
+   end
+end
+
+function block_parser.parse(input: Block, filename: string, parse_lang?: ParseLang): types.Node, {Error}, {string}
```
I saw that you omitted all error checking in this module: all those functions above are just returning `Node` and no errors, and are never checking that the sub-blocks they're expecting are present and have the correct kind.

I can imagine why you did that (because `teal.reader` already checks for errors). However, if we're planning to put a macro expansion step in between the reader and the parser, then the parser _does_ need to do complete error checking. 

In fact, the intention is that there will be _some_ parts of the error checking that we should be able to remove from the reader and have only in the parser. So the error checking will not be redundant. Basically:

* the "purely syntactic" checks will stay in the reader: e.g. ensuring that `for i = 10, 1, -1` is followed by a `do` token;
* all "structural" checks need to be present in the parser: e.g. that the blocks inside a `while` block are _really_ producing an expression sub-node and a "statements" sub-node.

To give you a concrete example as to why the current state without error checks is a problem: as it currently stands, without the checks, an invalid `while` block could just end up with `node.exp = nil` or `node.body = nil` and that would crash the type-checker later.

Frityet replied:
Yep I completely understand, I'll move it from the block reading to here

hishamhm replied:
Keep in mind that a lot of error checking does stay in the reader! It's not a problem if it's a little redundant at first; when in doubt, keep the checks in the reader when adding checks to the parser. It's better to have redundant checks at first than to have missing checks.

Frityet commented:
Will do, I always have so much trouble with Git 😅 never works how you think it should

Frityet committed:
add parser and reader again

I thought I had lost these guys! Git is scary...

Frityet committed:
Begin type parsing

hishamhm commented:
@Frityet you still haven't done this: 
https://github.com/Frityet/tl/pull/1#issuecomment-3005740347

Judging from your accidental force-push to _this branch_, looks like you did not run the first command from the three I sent you in that comment.

Frityet commented:
> @Frityet you still haven't done this: 
> 
> https://github.com/Frityet/tl/pull/1#issuecomment-3005740347
> 
> 
> 
> Judging from your accidental force-push to _this branch_, looks like you did not run the first command from the three I sent you in that comment.
> 
> 

I had ran that, after I pulled module-split from upstream, I had to force push because otherwise it wouldn't let me push but I think I might have done it wrong. Let me try it again

hishamhm commented on teal/block-parser.tl:
```diff
@@ -511,9 +585,14 @@ parse_fns.fornum = function(state: ParseState, block: Block): Node
+   local node = new_node(state, block, "forin")
    node.vars = parse_variable_list(state, block[1])
    node.exps = parse_expression_list(state, block[2])
+   if #node.exps < 1 then
```
What if `parse_expression_list` fails and returns `nil`?

Frityet replied:
`parse_expression_list` should never return nil, but the `#exps == 0` can happen, should I change this behaviour?

hishamhm replied:
> parse_expression_list should never return nil

If that's guaranteed, then this is fine. :+1:

hishamhm commented:
> I had ran that, after I pulled module-split from upstream

Ah, I see the confusion. That command was meant to create a new _local_ branch `module-split` (and then reset its commit to the one you originally based this branch on), not to fetch the current `module-split` from upstream (which has since changed).

>  I had to force push because otherwise it wouldn't let me push 

That is because you were not in your local `module-split` branch when you ran `git reset --hard`, but you were in your `block-parser` branch.

hishamhm commented:
Try again by creating a new branch with a unique name, called `old-module-split`:

```
git checkout master 
git checkout -b old-module-split
git reset --hard 586a137
git push
git checkout block-parser
```

And then retarget the PR to `old-module-split` by clicking the "Edit" button that should appear for you at the top right, next to "<> Code".

Frityet commented:
> Try again by creating a new branch with a unique name, called `old-module-split`:
> 
> ```
> git checkout master 
> git checkout -b old-module-split
> git reset --hard 586a137
> git push
> git checkout block-parser
> ```
> 
> And then retarget the PR to `old-module-split` by clicking the "Edit" button that should appear for you at the top right, next to "<> Code".

Ok I think that should be good? git is so confusing...

hishamhm commented:
That worked! You have now 2 changed files instead of 33.

Frityet commented:
Also, I am trying to add more comments (its a really bad habit of mine im trying to shake where I would usually almost never write comments), what do you think so far? Too "literal"/verbose?

hishamhm commented:
The branch now has only 2 files, but I believe you had more files (e.g. reader tests) previously. You should be able to restore those files by temporarily checking out locally your last commit before your accidental force-push: 52d4db213829d8b3a9f15956a64f1c330c7bfa77

Frityet commented:
> The branch now has only 2 files, but I believe you had more files (e.g. reader tests) previously. You should be able to restore those files by temporarily checking out locally your last commit before your accidental force-push: [52d4db2](https://github.com/Frityet/tl/commit/52d4db213829d8b3a9f15956a64f1c330c7bfa77)

Haha I knew I was missing something, thanks, will do

Frityet committed:
re-add reader tests

hishamhm committed:
refactor: major refactor of the compiler backend

hishamhm committed:
rename teal.checker -> teal.check

hishamhm committed:
rename tl.compat to tl.apply_compat

hishamhm committed:
refactor: move check() function to its own module

Frityet committed:
Add reader and block-parser to makefile

I was trying to debug my code for so long and was so confused why it seemed like none of my fixes was working... turns out I never put the sources in the makefile, and I was just running the lua code that I had generated manually...

Frityet committed:
Block-parser bug fixes + in depth unit-tests

hishamhm committed:
Makefile: allow using tl binary when bootstrapping

hishamhm committed:
extras: binary.sh updated for multi-module Teal

Also, we can now run the Makefile bootstrapping using
the Teal binary, like so:

```
extras/binary.sh
make STABLE_TL=_binary/build/tl
```

hishamhm committed:
Makefile: bin and revert targets

Useful for development.

hishamhm committed:
binary: quicker rebuild

hishamhm committed:
Makefile: more tweaks

hishamhm committed:
binary: ensure popen works

hishamhm committed:
mark lots of requires as abstract

mbartlett21 committed:
extras: add script to build a single-file tl module

Co-Authored-By: Hisham Muhammad <hisham@gobolinux.org>
